### ISSUE TITLE
"key" property missing from Flow object type aliases

### ISSUE THREAD

**Issue by Daniel15** [babel/babel-eslint#307]
Example code:

``` js
type MyType = {someProperty: Foo}
```

See http://astexplorer.net/#/n0uGG2VqNE

The `ObjectTypeProperty` should have a `key` property with the `someProperty` identifier and a `value` property with the `GenericTypeAnnotation`. However, the `key` is missing when using babel-eslint on ASTExplorer. Using babylon directly works fine.

**Comment by hzoo** [comment]
That's really odd..

**Comment by hzoo** [comment]
oh sigh lol https://github.com/babel/babel-eslint/blob/master/babylon-to-espree/toAST.js#L175-L178

git blame isn't helping since we moved the file int and out of the repo

**Comment by hzoo** [comment]
Ok https://github.com/babel/babel-eslint/pull/109/files

**Comment by Daniel15** [comment]
So it's actually intentional? Hmm, that's annoying, it makes it impossible to validate Flow object types when using babel-eslint. :disappointed:

**Comment by hzoo** [comment]
Sorry, I think I was just trying to make no-undef and no-unused-vars work for flow and those happened to be exceptions or something but obviously it shouldn't be there. I guess before it was just straight up removing the type altogether. Don't have time to look into this now though

**Comment by vjeux** [comment]
I just ran into a similar issue where the `key` attribute of `ClassProperty` is being removed (example http://astexplorer.net/#/4vTsbwrTFF ).

**Comment by pranaygp** [comment]
I ran into the same issue when writing `eslint-plugin-relay` and found a nice utility used by `eslint-plugin-flowtype`:

https://github.com/gajus/eslint-plugin-flowtype/blob/master/src/utilities/getParameterName.js#L20-L32

Hopefully useful to you if you stumble upon this issue and just want a workaround :)

**Comment by michaelhogg** [comment]
@pranaygp: Thank you, very nice workaround! :tada:

For the benefit of people coming from a search engine and looking for a quick solution, the [code](https://github.com/gajus/eslint-plugin-flowtype/blob/v2.41.0/src/utilities/getParameterName.js#L20-L32) is:

```js
if (node.type === 'ObjectTypeProperty') {

    let tokenIndex = 0;

    if (node.static) {
        tokenIndex++;
    }

    if (node.variance) {
        tokenIndex++;
    }

    const keyIdentifier = context.getSourceCode().getFirstToken(node, tokenIndex);

    // keyIdentifier has this structure:
    // {
    //     "type":  "Identifier",
    //     "value": "myProperty",
    //     "start": 827,
    //     "end":   837,
    //     "loc": {
    //         "start": {
    //             "line":   32,
    //             "column": 20
    //         },
    //         "end": {
    //             "line":   32,
    //             "column": 30
    //         }
    //     },
    //     "range": [
    //         827,
    //         837
    //     ]
    // }

}
```

The logic is:

:one: The node may be `static`, eg: [a `static` method for a class](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/static) in [a Flow libdef](https://flow.org/en/docs/libdefs/creation/#toc-declaring-a-global-class):

```js
declare class URL {
    static compare(url1: URL, url2: URL): boolean;
};
```

In this case, the first `Token` will be an `Identifier` with value `"static"`, so it needs to be skipped.

:two: The node may have a [`variance`](https://flow.org/en/docs/types/interfaces/#toc-interface-property-variance-read-only-and-write-only) operator, eg: `+` or `-`:

```js
type TypeDemo = {|
    +foo: number
|};
```

In this case, the first `Token` will be a `Punctuator` with value `"+"` or `"-"`, so it needs to be skipped.

:three: [`context.getSourceCode()`](https://eslint.org/docs/developer-guide/working-with-rules#the-context-object) returns a `SourceCode` object.

:four: [`.getFirstToken(node, skipOptions)`](https://eslint.org/docs/developer-guide/working-with-rules#contextgetsourcecode) skips the `static` and/or `variance` tokens if necessary, and returns a `Token` object of type `Identifier` which is equivalent to the missing `key`.

**Comment by kaicataldo** [comment]
Thank you for the issue. Now that [`@babel/eslint-parser` has been released](https://babeljs.io/blog/2020/07/13/the-state-of-babel-eslint), we are making this repository read-only. If this is a change you would still like to advocate for, please reopen this in the [babel/babel monorepo](https://github.com/babel/babel).
